providers/obs/obs.go

```diff
@@ -0,0 +1,416 @@
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "fail to multipart upload")
+		}
+		parts = append(parts, obs.Part{PartNumber: output.PartNumber, ETag: output.ETag})
+	}
+	return parts, nil
+}
+
+func (b *Bucket) Close() error { return nil }
+
+// Iter calls f for each entry in the given directory (not recursive.)
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error, options ...objstore.IterOption) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, DirDelim) + DirDelim
+	}
+
+	input := &obs.ListObjectsInput{}
+	input.Bucket = b.name
+	input.Prefix = dir
+	input.Delimiter = DirDelim
+	if objstore.ApplyIterOptions(options...).Recursive {
+		input.Delimiter = ""
+	}
+	for {
+		output, err := b.client.ListObjects(input)
+		if err != nil {
+			return errors.Wrap(err, "fail to list object")
+		}
+		for _, content := range output.Contents {
+			if err := f(content.Key); err != nil {
+				return errors.Wrap(err, "fail to call f for object")
+			}
+		}
+		for _, topDir := range output.CommonPrefixes {
+			if err := f(topDir); err != nil {
+				return errors.Wrap(err, "fail to call f for top dir object")
+			}
+		}
+
+		if !output.IsTruncated {
+			break
+		}
+
+		input.Marker = output.NextMarker
+	}
+	return nil
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, 0, -1)
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, off, length)
+}
+
+func (b *Bucket) getRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	if strings.TrimSpace(name) == "" {
+		return nil, errors.New("object name cannot be empty")
+	}
+	input := &obs.GetObjectInput{}
+	input.Bucket = b.name
+	input.Key = name
+	if off < 0 {
+		return nil, errors.New("incorrect offset")
+	}
+	input.RangeStart = off
+	if length != -1 {
+		input.RangeEnd = off + length - 1
+	} else {
+		input.RangeEnd = math.MaxInt64
+	}
+	output, err := b.client.GetObject(input)
+	if err != nil {
+		return nil, errors.Wrap(err, "fail to get object")
+	}
+	return output.Body, nil
+}
+
+// Exists checks if the given object exists in the bucket.
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	_, err := b.client.GetObjectMetadata(&obs.GetObjectMetadataInput{
+		Bucket: b.name,
+		Key:    name,
+	})
+	if err != nil {
+		if b.IsObjNotFoundErr(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "fail to get object metadata")
+	}
+	return true, nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	switch oriErr := errors.Cause(err).(type) {
+	case obs.ObsError:
+		if oriErr.Status == "404 Not Found" {
```

Comment:
Since we have only one case, why not do `oriErr, ok := errors.Cause(err).(obs.ObsError)`?